tracer/src/Datadog.Trace/Activity/ActivityListener.cs

```diff
@@ -0,0 +1,511 @@
+        private static Action<string, KeyValuePair<string, object>, object> _onNextActivityDelegate;
+        private static Func<object, bool> _onSetListenerDelegate;
+
+        private static object _activityListenerInstance;
+
+        private static int _initialized = 0;
+        private static int _stopped = 0;
+
+        public static IActivity GetCurrentActivity()
+        {
+            var activity = _getCurrentActivity?.Invoke();
+            if (activity is null)
+            {
+                return null;
+            }
+
+            if (activity.TryDuckCast<IActivity6>(out var activity6))
+            {
+                return activity6;
+            }
+
+            if (activity.TryDuckCast<IActivity5>(out var activity5))
+            {
+                return activity5;
+            }
+
+            if (activity.TryDuckCast<IW3CActivity>(out var w3cActivity))
+            {
+                return w3cActivity;
+            }
+
+            return activity.TryDuckCast<IActivity>(out var activity4) ? activity4 : null;
+        }
+
+        public static void Initialize() => Initialize(CancellationToken.None);
+
+        public static void Initialize(CancellationToken cancellationToken)
+        {
+            if (Interlocked.CompareExchange(ref _initialized, 1, 0) == 1)
+            {
+                return;
+            }
+
+            // Try to resolve System.Diagnostics.DiagnosticListener, System.Diagnostics.DiagnosticSource
+            _diagnosticListenerType = Type.GetType("System.Diagnostics.DiagnosticListener, System.Diagnostics.DiagnosticSource");
+            if (_diagnosticListenerType is null)
+            {
+                // Because we cannot resolve the DiagnosticListener type we allow a later initialization.
+                // For this case we are going to do some retries with a back-off.
+                if (Interlocked.Decrement(ref _initializationRetries) > 0)
+                {
+                    Task.Delay(InitializationBackoffPerRetry, cancellationToken).ContinueWith(
+                        _ =>
+                    {
+                        if (cancellationToken.IsCancellationRequested)
+                        {
+                            return;
+                        }
+
+                        Interlocked.Exchange(ref _initialized, 0);
+                        Initialize();
+                    },
+                        cancellationToken);
+                }
+
+                return;
+            }
+
+            // Initialize
+            var diagnosticSourceAssemblyName = _diagnosticListenerType.Assembly.GetName();
+            Log.Information("DiagnosticSource: {diagnosticSourceAssemblyNameFullName}", diagnosticSourceAssemblyName.FullName);
+
+            var version = diagnosticSourceAssemblyName.Version;
+
+            // Check if Version 5 or 6 is loaded (Uses ActivityListener implementation)
+            if (version?.Major is 5 or 6)
```

Comment:
Can we do a `>= 5` check instead of `is 5 or 6`?

Reply:
done in https://github.com/DataDog/dd-trace-dotnet/pull/2446/commits/9c00bdcbc3c6acb0412def01171646c53701314a